apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResourceTest.java

```diff
@@ -0,0 +1,221 @@
+                    ClickHouseContainerUtils.migrationParameters());
+        }
+
+        this.baseURI = "http://localhost:%d".formatted(client.getPort());
+        this.client = client;
+        this.mySqlTemplate = mySqlTemplate;
+
+        ClientSupportUtils.config(client);
+    }
+
+    private static void mockTargetWorkspace(String apiKey, String workspaceName, String workspaceId) {
+        AuthTestUtils.mockTargetWorkspace(wireMock.server(), apiKey, workspaceName, workspaceId, USER);
+    }
+
+    @AfterAll
+    void tearDownAll() {
+        wireMock.server().stop();
+    }
+
+    @Test
+    @DisplayName("Create and update provider Api Key")
+    void createAndUpdateProviderApiKey() {
+
+        String workspaceName = UUID.randomUUID().toString();
+        String apiKey = UUID.randomUUID().toString();
+        String workspaceId = UUID.randomUUID().toString();
+        var provider = OPEN_AI;
+        String providerApiKey = factory.manufacturePojo(String.class);
+
+        mockTargetWorkspace(apiKey, workspaceName, workspaceId);
+
+        var id = createProviderApiKey(provider, providerApiKey, apiKey, workspaceName, 201);
+        var expectedProviderApiKey = ProviderApiKey.builder().id(id).provider(provider).build();
+        getAndAssertProviderApiKey(expectedProviderApiKey, apiKey, workspaceName);
+        checkEncryption(id, workspaceId, providerApiKey);
+
+        String newProviderApiKey = factory.manufacturePojo(String.class);
+        updateProviderApiKey(id, newProviderApiKey, apiKey, workspaceName, 204);
+        checkEncryption(id, workspaceId, newProviderApiKey);
+    }
+
+    @Test
+    @DisplayName("Create provider Api Key for existing provider should fail")
+    void createProviderApiKeyForExistingProviderShouldFail() {
+
+        String workspaceName = UUID.randomUUID().toString();
+        String apiKey = UUID.randomUUID().toString();
+        String workspaceId = UUID.randomUUID().toString();
+        var provider = OPEN_AI;
+        String providerApiKey = factory.manufacturePojo(String.class);
+
+        mockTargetWorkspace(apiKey, workspaceName, workspaceId);
+
+        createProviderApiKey(provider, providerApiKey, apiKey, workspaceName, 201);
+        createProviderApiKey(provider, providerApiKey, apiKey, workspaceName, 409);
+    }
+
+    @Test
+    @DisplayName("Update provider Api Key for non-existing Id")
+    void updateProviderFail() {
+
+        String workspaceName = UUID.randomUUID().toString();
+        String apiKey = UUID.randomUUID().toString();
+        String workspaceId = UUID.randomUUID().toString();
+        String providerApiKey = factory.manufacturePojo(String.class);
+
+        mockTargetWorkspace(apiKey, workspaceName, workspaceId);
+
+        // for non-existing id
+        updateProviderApiKey(UUID.randomUUID(), providerApiKey, apiKey, workspaceName, 404);
+    }
+
+    private UUID createProviderApiKey(LlmProvider provider, String providerApiKey, String apiKey, String workspaceName,
```

Comment:
We should use Podam to generate the beans. Also, we can extract these cans to a ResourceClient class like `SpanResourceClient,` for reusability.

Reply:
Agree about reusability. Regarding Podam, `LlmProvider ` can be generated within the method, but `providerApiKey` should be passed as param, since it's used lated in the outer scope.